### PR TITLE
[apex] New Rule - Test Methods Must Be In Test Classes

### LINKED ISSUE #639: [apex] Test methods should not be in classes other than test classes
**Rule Set:** Apex Unit

**Description:**
Methods marked as a testMethod but not residing in a test class should be movedto a proper class or should add the "@isTest" annotation to the class

**Code Sample demonstrating the issue:**

```
public class myClass {
static testMethod void myTest() {} //missing static keyword
}
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

### PR DESCRIPTION
**Pull request by noerremark**
Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
The title pretty must says what this is for. It fixes #639.

### CODE REVIEW

**Comment by rsoesemann** [comment]
Sorry to be the bad cop here but I think this is a rule with really little value. Everybody should use the annotation and not the keyword for years. The annotation can only be added to @istest classes.

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by rsoesemann** [comment]
@noerremark what about https://github.com/pmd/pmd/issues/2210?!

**Comment by noerremark** [comment]
> Sorry to be the bad cop here but I think this is a rule with really little value. Everybody should use the annotation and not the keyword for years.

I did this because #639 was open, and felt like something easy to start on. But we can close #639 and abandon this PR too, if you like.

I don't necessarily disagree with the bad cop in you, but I DO think that "the Salesforce"-way justifies a rule like this for the following reasons:

1. I don't get a warning in Illuminated Cloud if I remove @IsTest from a class with tests. Having this in place may mean that I get that without building.
2. Since building happens Salesforce-side, it will always be slower than running PMD on local

Re: `testMethod` is still heavily [used in examples in documentation](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation_isTest.htm), and I wouldn't be surprised if lots of people out there still uses it.

**Comment by dstdia** [comment]
There are two sides to this rule (and the general idea): In current API versions, the Apex Compiler will reject test method annotations if the class itself is not annotated with @isTest. So you won't need any PMD rule because the compiler will find this at compile time.

The other aspect is: i'm reviewing implementations or I'm taking over dated codebases from time to time, and one of the first things that I do is running Apex PMD against the codebase. Finding code that still uses ancient API versions that allow for @testMethod within regular classes (or even "require" these inline tests) would be much easier if there was a rule pointing at these classes. 

You may argue if this is the primary focus of Apex PMD, and you probably don't need such a rule in a basic / general ruleset, but I think it's handy for retro-analysis.

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/TestMethodsMustBeInTestClasses [hunk 1]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (adangel) [inline pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/TestMethodsMustBeInTestClasses]: This file should have a `xml` extension

Comment (adangel) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/TestMethodsMustBeInTestClasses]: And we need a JUnit Test class to start the tests - right now, the new rule is not tested at all....
Example: https://github.com/pmd/pmd/blob/master/pmd-apex/src/test/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/ApexUnitTestClassShouldHaveAssertsTest.java
The test class is pretty empty, but still required

Docu: https://pmd.github.io/latest/pmd_userdocs_extending_testing.html

Comment (dstdia) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/TestMethodsMustBeInTestClasses]: For the sake of completeness: API version 27.0 (Spring '13) removed support for test methods inside of functional classes. I didn't realize that was so long ago, so probably the relevance even in legacy code is getting smaller and smaller...

**Comment by noerremark** [comment]
@adangel / @dstdia / @rsoesemann - Better now?

**Comment by rsoesemann** [comment]
Let’s add it. It’s just a simple XPath rule so not much code everhead. This is not a rule we need to make pmd more attractive but it’s also doesn’t do any harm.

**Comment by noerremark** [comment]
@adangel - What is the rule, can I merge it, or will you?

**Comment by adangel** [comment]
@noerremark  I'll merge it

**Comment by adangel** [comment]
Merged - this will be part of PMD 6.22.0

**Pull request merged**